Add --broker-namespace flag to deploy-broker

donaldh opened the pull request:
Deploys broker to custom namespace. Submariner join uses broker namespace
from broker-info.subm

Signed-off-by: Donald Hunter <donald.hunter@redhat.com>

Implemented as a first step to supporting multiple "tenant" brokers in a single cluster and could be part of what is needed for #1040 – Support deployment within a custom namespace.

skitt commented on pkg/broker/ensure.go:
```diff
@@ -74,88 +74,88 @@ func Ensure(config *rest.Config, componentArr []string, crds bool) error {
 
 // CreateSAForCluster creates a new SA, and binds it to the submariner cluster role
-func CreateSAForCluster(clientset *kubernetes.Clientset, clusterID string) (*v1.Secret, error) {
+func CreateSAForCluster(clientset *kubernetes.Clientset, clusterID string, namespace string) (*v1.Secret, error) {
```
```suggestion
func CreateSAForCluster(clientset *kubernetes.Clientset, clusterID, namespace string) (*v1.Secret, error) {
```

skitt commented on pkg/broker/ensure.go:
```diff
@@ -182,25 +182,25 @@ func WaitForClientToken(clientset *kubernetes.Clientset, submarinerBrokerSA stri
 }
 
-func CreateNewBrokerRoleBinding(clientset *kubernetes.Clientset, serviceAccount, role string) (brokerRoleBinding *rbac.RoleBinding,
+func CreateNewBrokerRoleBinding(clientset *kubernetes.Clientset, serviceAccount, role string, namespace string) (brokerRoleBinding *rbac.RoleBinding,
```
```suggestion
func CreateNewBrokerRoleBinding(clientset *kubernetes.Clientset, serviceAccount, role, namespace string) (brokerRoleBinding *rbac.RoleBinding,
```

skitt commented on pkg/broker/rbac.go:
```diff
@@ -110,7 +110,7 @@ func NewBrokerClusterRole() *rbacv1.Role {
 
 // Create a role for to bind the cluster admin (subctl) SA
-func NewBrokerRoleBinding(serviceAccount, role string) *rbacv1.RoleBinding {
+func NewBrokerRoleBinding(serviceAccount, role string, namespace string) *rbacv1.RoleBinding {
```
```suggestion
func NewBrokerRoleBinding(serviceAccount, role, namespace string) *rbacv1.RoleBinding {
```

skitt commented on pkg/subctl/cmd/root.go:
```diff
@@ -112,16 +112,27 @@ func handleNodeLabels(config *rest.Config) error {
 		}
 		if answer.Node == "" {
 			fmt.Printf("* No worker node found to label as the gateway\n")
+			nodes, err := clientset.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+			if err != nil {
+				return err
+			}
+			if len(nodes.Items) == 0 {
+				return fmt.Errorf("There are no nodes in the cluster")
+			}
+			node = nodes.Items[0].GetName()
+			fmt.Printf("* Labelling %s as gateway\n", node)
```
This seems unrelated to the broker namespace change, or am I missing something?

donaldh replied:
Apologies, this should not be part of the commit. Is there an easy way to remove this from the PR, or should I submit a new PR?

skitt replied:
You can amend your commit and force push your branch; you’ll need to do that anyway because the additional commit generated by GitHub with my suggestions doesn’t carry a DCO... If you’d like help with the git commands feel free to ask.

donaldh replied:
Thx, that worked.

skitt commented on pkg/subctl/cmd/deploybroker.go:
```diff
@@ -73,6 +74,8 @@ func init() {
 
 	deployBroker.PersistentFlags().BoolVar(&operatorDebug, "operator-debug", false, "enable operator debugging (verbose logging)")
 
+	deployBroker.PersistentFlags().StringVar(&brokerNamespace, "broker-namespace", broker.SubmarinerBrokerNamespace, "namespace for broker")
```
I think we should move `broker.SubmarinerBrokerNamespace` here, and make it a tool-specific default. We can pretty much get rid of `pkg/broker/namespace.go` entirely given that `NewBrokerNamespace()` is trivial (it can be merged in `CreateNewBrokerNamespace()`).

skitt reviewed (changes requested):
Thanks for submitting this!

I have a few changes I’d like to see, mostly stylistic (repeated parameter types). There seems to be an unrelated change re. node labels.

donaldh committed:
Add --broker-namespace flag to deploy-broker

Deploys broker to custom namespace. Submariner join uses broker namespace
from broker-info.subm

Signed-off-by: Donald Hunter <donald.hunter@redhat.com>

dfarrell07 reviewed (approved):
LGTM, thanks! The linter's complaints:

```
pkg/broker/ensure.go:158:1: paramTypeCombine: func(clientset *kubernetes.Clientset, submarinerBrokerSA string, namespace string) (secret *v1.Secret, err error) could be replaced with func(clientset *kubernetes.Clientset, submarinerBrokerSA, namespace string) (secret *v1.Secret, err error) (gocritic)
func WaitForClientToken(clientset *kubernetes.Clientset, submarinerBrokerSA string, namespace string) (secret *v1.Secret, err error) {
^
pkg/broker/ensure.go:209:1: paramTypeCombine: func(clientset *kubernetes.Clientset, submarinerBrokerSA string, namespace string) (brokerSA *v1.ServiceAccount, err error) could be replaced with func(clientset *kubernetes.Clientset, submarinerBrokerSA, namespace string) (brokerSA *v1.ServiceAccount, err error) (gocritic)
func CreateNewBrokerSA(clientset *kubernetes.Clientset, submarinerBrokerSA string, namespace string) (brokerSA *v1.ServiceAccount, err error) {
^
pkg/subctl/operator/brokercr/ensure.go:41: File is not `gofmt`-ed with `-s` (gofmt)
```

tpantelis commented on pkg/subctl/cmd/deploybroker.go:
```diff
@@ -38,6 +38,10 @@ import (
 )
 
+const (
+	DefaultBrokerNamespace = "submariner-k8s-broker"
```
This doesn't need to be exported so let's not leak it.

```suggestion
	defaultBrokerNamespace = "submariner-k8s-broker"
```

donaldh committed:
fixup! Add --broker-namespace flag to deploy-broker

Resolve review comments

Signed-off-by: Donald Hunter <donald.hunter@redhat.com>

nyechiel commented:
@donaldh @dfarrell07 can you please ensure the new flag is documented?

donaldh commented:
Hi Nir,

Yes, I will do that. It will likely be January before I get to it. On holiday now. 

Thanks,
Donald

> On 22 Dec 2021, at 11:00, Nir Yechiel ***@***.***> wrote:
> 
> ﻿
> @donaldh @dfarrell07 can you please ensure the new flag is documented?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you were mentioned.

nyechiel commented:
Thanks @donaldh. I reported this issue so that we don't forget about this. Feel free to assign to yourself (I think it can only be assigned to you after you add a comment there): https://github.com/submariner-io/submariner-website/issues/630

Enjoy the break!

mkolesnik commented on pkg/subctl/cmd/deploybroker.go:
```diff
@@ -73,6 +78,8 @@ func init() {
 
 	deployBroker.PersistentFlags().BoolVar(&operatorDebug, "operator-debug", false, "enable operator debugging (verbose logging)")
 
+	deployBroker.PersistentFlags().StringVar(&brokerNamespace, "broker-namespace", defaultBrokerNamespace, "namespace for broker")
```
I know this was merged a while ago, but isn't it simpler UX and more consistent to just name this "--namespace"?

I'm talking about the flag name not the variable name.

donaldh replied:
The name was chosen to make it clear that it sets the broker namespace, as opposed to the operator namespace, or more generally to avoid suggesting the command's effect would only apply to a single `--namespace`

mkolesnik replied:
Yes that might make sense, but considering we're talking about `subctl deploy-broker` command I think having `broker` in the flag as well is repetitive.

Also, aren't we essentially affecting a single namespace here? So I'm not sure it's incorrect in that regard.
